doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,601 @@
+NEP 50 — Promotion rules for Python scalars
+===========================================
+:Author: Sebastian Berg
+:Status: Draft
+:Type: Standards Track
+:Created: 2021-01-10
+
+
+Abstract
+========
+
+Since NumPy 1.7 promotion rules are defined through the "safe casting"
+concept which in turn can depend on the actual values involved.
+While these choices were well intended, they lead to a complexity both
+for users and maintainers.
+
+There are two main ways this can lead to confusing results:
+
+1. Value=based promotion means that the value can matter::
+
+     np.result_type(np.int8, 1) == np.int8
+     np.result_type(np.int8, 255) == np.int16
+
+   which is even true when replacing the value with 0-D arrays::
+
+     int64_0d_array = np.array(1, dtype=np.int64)
+     np.result_type(np.int8, int64_0d_array) == np.int8
+    
+   This logic arises, because ``1`` can be represented by an ``int8`` while
+   ``255`` can be represented by an ``int16`` *or* ``uint8``.
+   Because of this, the exact type is often ignored for 0-D arrays or
+   NumPy scalars.
+
+2. For a Python ``int``, ``float``, or ``complex`` the value is inspected as
+   before.  But often surprisingly not when the NumPy object is a 0-D array
+   or NumPy scalar::
+
+     np.result_type(np.array(1, dtype=np.uint8), 1) == np.int64
+     np.result_type(np.int8(1), 1) == np.int64
+
+   The reason is that the special value-based promotion is disabled when all
+   objects are scalars or 0-D arrays.
+   NumPy thus returns the same type as ``np.array(1)``, which is usually
+   an ``int64`` (this depends on the system).
+
+Note that the examples apply also to operations like multiplication,
+addition, or comparisons and the corresponding functions like `np.multiply`.
+
+This NEP proposes to refactor the behaviour around two guiding principles:
+
+1. The value must never influence the result type.
+2. NumPy scalars or 0-D arrays must always lead to the same behaviour as
+   their N-D counterparts.
+
+We propose to remove all value-based logic and add special handling for
+Python scalars to preserve some of the convenience that it provided.
+This will mean that Python scalars are considered "weakly" typed.
+A Python integer will always be converted to the NumPy dtype retaining the
+behaviour that::
+
+    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
+    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array
+
+but removing any dependence on the Python value itself.
+
+The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
+we expect that the behaviour in functions (promotion) will in practice be far
+more important than the casting change itself.
+
+
+Schema of the new proposed promotion rules
+------------------------------------------
+
+After the change, the promotions in NumPy will follow the schema below.
+Promotion always occurs along the green lines:
+from left to right within their kind and to a higher kind only when
```

Comment:
You are right, will move the bubble.  It is super annoying, because technically, the bubble should be to the left of `uint64/int64`, but for practical reasons, numpy chooses to not put it there.
But... moving the bubble is right in practice for that reason.